Breaking changes to integrations mitigation process (#772)
OpenAssetIO/OpenAssetIO: doc/contributing/PROCESS.md

  uploaded correctly.
- You're done! Take a break and relax!

> ## Breaking Integrations

Reviewer: [consider][minor] We have a bit of a mixture of capitalisation style for titles in this doc. Mostly we use "Capitalise first word only".  Would be good to be consistent.

Author: Fixed the rest of the document as well, there were a couple. [a0b6406](https://github.com/OpenAssetIO/OpenAssetIO/pull/772/commits/a0b6406421d8641b3ef902737ba9847e55ad5ff9)

OpenAssetIO is pursuing an integration heavy validation strategy, it
seeks to have a set of robust proof of concept projects, and will use
these as both design validators and testing coverage providers.

In order to tie this coverage into mainline OpenAssetIO development,
OpenAssetIO continuous integration checks out several downstream
projects each time a commit is pushed. These projects are built against
the bleeding edge version of OpenAssetIO, and their test suites are run.
This provides valuable end to end testing of each changeset, and is
invaluable. (See

> [integrations.yml](https://github.com/OpenAssetIO/OpenAssetIO/blob/main/.github/workflows/integrations.yml))

Reviewer: [consider] Using full URLs is more portable on the web, but if we assume we're not going to host this .md file elsewhere, then it's better to use relative paths.  

E.g. my IDE can follow relative paths in the docs when I view them locally in a markdown preview.

Author: Hmm, guess this was workflow bias, i tend to view code in the browser quite a lot even when i've got it available locally, will change over to relative paths.

Reviewer 2: :pray:  selfishly, this serves me better as I often use the link to "open the right file to edit" (for this, and more generally in process documentation).

using a released OpenAssetIO version, taken from PyPI.

### Breaking Change Mitigation Process

On occasion, it is necessary for OpenAssetIO to change its interfaces in
such a way that cause the downstream projects to become incompatible
against the bleeding edge OpenAssetIO. You will know when this happens
as one of the tests in
[integrations.yml](https://github.com/OpenAssetIO/OpenAssetIO/blob/main/.github/workflows/integrations.yml)
will begin to fail.

The downstream projects are pinned to a specific version of OpenAssetIO
on PyPI for their own test suites, so these will continue to function
even through a breaking change on OpenAssetIO main. However, the
OpenAssetIO integrations checks will being to fail, with no clear way to
remedy the situation before an OpenAssetIO release, as any changes to
the downstream project to support bleeding edge will likely cause
incompatibility with the published OpenAssetIO release the project is
consuming from PyPI.

In order to remedy this, when making a breaking OpenAssetIO change that
causes a failing integrations check, the following procedure may be
implemented to avoid long term loss of coverage.

> - Consult on whether the change must necessarily be breaking, if

Reviewer: [consider] Use numeric list, rather than bullets, since it's a fixed sequence (and opens up the possibility of cross-referencing steps).

Author: [44a9b4b](https://github.com/OpenAssetIO/OpenAssetIO/pull/772/commits/44a9b4b40357ae7eb8ee576cd93afb7b3d644e21)

> - Consult on whether the change must necessarily be breaking, if
> reasonably possible, attempt to formulate another approach.

Reviewer: [consider] For consistency we should indent the next line(s) of bullet points
```suggestion
- Consult on whether the change must necessarily be breaking, if
  reasonably possible, attempt to formulate another approach.
```

- Evaluate the scope of breakage in the dependent project(s), and make
an issue in those projects to fix compatibility with your change.

- Comment out the failing integrations check in
[integrations.yml](https://github.com/OpenAssetIO/OpenAssetIO/blob/main/.github/workflows/integrations.yml).

> Leave a note explaining the context, with links to the relevant issues.

Reviewer: [consider] Be more specific?
```suggestion
Leave a note explaining the context, with a link to the issue created in
the previous step.
```

- Review and merge your change to OpenAssetIO as normal, call out the
removed integrations check in the PR.

> - As soon as possible, plan to work on the issues to return
> compatibility to the dependent repos.

Reviewer: [consider] Be more specific?

```suggestion
- As soon as possible, plan to work on the issue(s) created in step 2 to 
return compatibility to the dependent repos.
```

Author: [44a9b4b](https://github.com/OpenAssetIO/OpenAssetIO/pull/772/commits/44a9b4b40357ae7eb8ee576cd93afb7b3d644e21)

- To accomplish this, branch the dependent repo in question from main,

> call your new branch `openassetio/X.Y.Z`

Reviewer: [consider]
```suggestion
call your new branch `openassetio/X.Y.Z`, where `X.Y.Z` is the _next_
release of OpenAssetIO.
```

Author: (Commited a few fixes together, soz) [04f2813](https://github.com/OpenAssetIO/OpenAssetIO/pull/772/commits/04f28134a9dd0e08236ba9a9076f499bd9d4dfff)

  > **Warning**
  >

>   > The specific branch name formatting is important, as CI is setup to

Reviewer: [fix] typo
```suggestion
  > The specific branch name formatting is important, as CI is set up to
```

Author: (Commited a few fixes together, soz) [04f2813](https://github.com/OpenAssetIO/OpenAssetIO/pull/772/commits/04f28134a9dd0e08236ba9a9076f499bd9d4dfff)

  > source wheels built from the tip of `openassetio:main` when run on
  > these branches

- Fix the project to function with the new interface. Review and Merge
to the `openassetio/X.Y.Z` branch.

- Re-enable the disabled integration tests in OpenAssetIO by re-targeting
the checkout in
[integrations.yml](https://github.com/OpenAssetIO/OpenAssetIO/blob/main/.github/workflows/integrations.yml)
to point to this new branch. It should now pass.

> - When OpenAssetIO has its next release, fold these branches back into
>   the main branches and delete them.
> 
> - Re-target
> [integrations.yml](https://github.com/OpenAssetIO/OpenAssetIO/blob/main/.github/workflows/integrations.yml)
> back to normal, checking out the `main` branch of the dependent repos.

Reviewer: [consider] I know it should be pretty obvious, but to be safe we shouldn't delete the staging branch until there's no chance of it being used.

```suggestion
- When OpenAssetIO has its next release, fold these branches back into
  the main branches.

- Re-target
[integrations.yml](https://github.com/OpenAssetIO/OpenAssetIO/blob/main/.github/workflows/integrations.yml)
back to normal, checking out the `main` branch of the dependent repos.

- Delete the `openassetio/X.Y.Z` staging branch.
```

Author: Really good call imo, [9d04a58](https://github.com/OpenAssetIO/OpenAssetIO/pull/772/commits/9d04a58afb9f23a32f2d16986eecc982a1949b68)

causes a failing integrations check, the following procedure may be
implemented to avoid long term loss of coverage.

1. Consult on whether the change must necessarily be breaking, if
   reasonably possible, attempt to formulate another approach.

2. Evaluate the scope of breakage in the dependent project(s), and make
   an issue in those projects to fix compatibility with your change.

3. Comment out the failing integrations check in
   [integrations.yml](../../.github/workflows/integrations.yml). Leave a
   note explaining the context, with links to the issues created in the
   previous step.

4. Review and merge your change to OpenAssetIO as normal, call out the
   removed integrations check in the PR.

5. As soon as possible, plan to work on the issues(s) created in step 2
   to return compatibility to the dependent repos.

6. To accomplish this, branch the dependent repo in question from main,
   call your new branch `openassetio/X.Y.Z`, where `X.Y.Z` is the *next*
   release of OpenAssetIO.

   > **Warning**
   >
   > The specific branch name formatting is important, as CI is setup to
   > source wheels built from the tip of `openassetio:main` when run on

>    > these  branches

Reviewer 2: [consider] I'm not sure quite where to put this, but it feels like we should highlight that a downstream project _may_ not have CI already configured to pull `main` wheels for dep feature branches as described.

I imagine that the first time we need to go through this process, we'll add the CI tweaks, vs up front.

Author: Yeah, I was writing this document from the perspective of the future, in the hopes that it would become true in time and we wouldn't have to muddy the document, as it's already quite complex.

However, as it's possible we won't get to doing this on any specific repo for a while, I'll add a clarifying sentence.

Author: [3fb73ab](https://github.com/OpenAssetIO/OpenAssetIO/pull/772/commits/3fb73ab95cc522add5fb7e1159467e8d3d797758)
